ag-master/worker.py

```diff
@@ -1,74 +1,66 @@
+            if job.status not in at_list:
+                raise PermissionError
+            try:
+                return func(job=job, **kwargs)
+            finally:
+                job.status = to
+                db.session.commit()
+
+        return handler
 
+    return decorator
+
+
+def create_worker_endpoints(app):
     @get_submission.bind(app)
-    @check_course_secret
-    def get_submission_rpc(course, bid, job_id):
-        job = Job.query.filter_by(job_key=job_id).first()
-        assignment = Assignment.query.filter_by(
-            ag_key=job.assignment, course=course.secret
-        ).first()  # validates secret
-        if job and assignment:
-            r = requests.get(
-                SUBM_ENDPOINT + "/" + str(bid),
-                params=dict(access_token=job.access_token),
-            )
-            r.raise_for_status()
-            return r.json()["data"]
-        return dict(success=False)
+    @job_transition(at="queued", to="started")
+    def get_submission_rpc(job):
+        r = requests.get(
+            f"{SUBM_ENDPOINT}/{job.backup}",
+            params=dict(access_token=job.access_token),
+        )
+        r.raise_for_status()
+        return r.json()["data"]
 
     @handle_output.bind(app)
-    @check_course_secret
-    def handle_output_rpc(course, output, job_id):
-        job = Job.query.filter_by(job_key=job_id).first()
-        assignment = Assignment.query.filter_by(
-            ag_key=job.assignment, course=course.secret
-        ).first()  # validates secret
-        if job and assignment:
-            scores = parse_scores(output)
-            for score in scores:
-                score["bid"] = job.backup
-                requests.post(
-                    SCORE_ENDPOINT,
-                    data=score,
-                    params=dict(access_token=job.access_token),
-                )
-        return dict(success=(job is not None))
+    @job_transition(at="started", to="finished")
+    def handle_output_rpc(output, job):
+        scores = parse_scores(output)
+        for score in scores:
+            score["bid"] = job.backup
+            requests.post(
+                SCORE_ENDPOINT,
+                data=score,
+                params=dict(access_token=job.access_token),
+            )
 
     @set_results.bind(app)
-    @check_course_secret
-    def set_results_rpc(course, job_id, status, result):
-        job = Job.query.filter_by(job_key=job_id).first()
-        assignment = Assignment.query.filter_by(
-            ag_key=job.assignment, course=course.secret
-        ).first()  # validates secret
-        if job and assignment:
-            job.status = status
-            job.result = result
-            db.session.commit()
-        return dict(success=(job is not None))
+    @job_transition(at=["queued", "started"], to="failed")
+    def set_results_rpc(job, status, result):
+        job.status = status
+        job.result = result
+        db.session.commit()
```

Comment:
this should be changed to `set_failure` now, since whatever is passed to status is redundant